add total resources requested to podstatus

tzneal opened the pull request:
#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
/kind feature

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

#### Special notes for your reviewer:


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
Yes
```release-note
The effective resource requests for pods are now reported on their status via a new TotalResourcesRequested field and in the output of kubectl describe pod.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/0dc6db753668b2e7fc515d82963abbe0ba1a79f4/keps/sig-node/753-sidecar-containers/README.md
```



Sample output of kubectl describe pod:
```
<--cut-->
Init Containers:
  alpine:
    Image:      alpine
    Port:       <none>
    Host Port:  <none>
    Command:
      sleep
      5
    Requests:
      cpu:        9500m
    Environment:  <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-nzbq9 (ro)
Containers:
  nginx:
    Image:      nginx:1.14.2
    Port:       80/TCP
    Host Port:  0/TCP
    Requests:
      cpu:        100m
      memory:     70Mi
    Environment:  <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-nzbq9 (ro)
Conditions:
  Type           Status
  PodScheduled   False
Volumes:
  kube-api-access-nzbq9:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Total Resources Requested:
  cpu:           9500m
  memory:        70Mi
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason            Age   From               Message
  ----     ------            ----  ----               -------
  Warning  FailedScheduling  10s   default-scheduler  0/1 nodes are available: 1 Insufficient cpu. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod..
```

SergeyKanzhelev commented:
/ok-to-test

SergeyKanzhelev commented:
/priority important-soon
/triage accepted

SergeyKanzhelev commented on pkg/apis/core/types.go:
```diff
@@ -3523,6 +3523,11 @@ type PodStatus struct {
 	// +optional
 	EphemeralContainerStatuses []ContainerStatus
+
+	// TotalResourcesRequested is the effective resource requirements for this pod, taking into consideration init containers, sidecars, containers and pod overhead.
```
lets not mentione sidecars before they are implemented

```suggestion
	// TotalResourcesRequested is the effective resource requirements for this pod, taking into consideration init and regular containers, and the pod overhead.
```

SergeyKanzhelev replied:
please explain how this behaves if request is not set

SergeyKanzhelev commented on pkg/scheduler/schedule_one.go:
```diff
@@ -946,7 +946,11 @@ func updatePod(ctx context.Context, client clientset.Interface, pod *v1.Pod, con
+	if !condNeedsUpdate && !nnnNeedsUpdate && !totalResourcesNeedsUpdate {
 		return nil
 	}
 	if nnnNeedsUpdate {
```
nit: maybe this can follow the pattern and actually make an update before the previous `if` condition.

SergeyKanzhelev reviewed (approved):
/lgtm

tzneal commented:
/retest

SergeyKanzhelev commented on pkg/apis/core/types.go:
```diff
@@ -3595,6 +3595,13 @@ type PodStatus struct {
 	// +optional
 	Resize PodResizeStatus
+
+	// TotalResourcesRequested is the effective resource requests for this pod taking into consideration init
```
this will include the requested for InPlace resize, correct? Maybe we can mention it here?

tzneal committed:
api: add TotalRequestedResources to PodStatus

- api: add RequestedResourcesField
- kube-scheduler: update RequestedResources upon failure to schedule
- kubelet: add RequestedResources on pod status update
- kubectl: show RequestedResources in kubectl describe pod

tzneal commented on pkg/api/v1/pod/util.go:
```diff
@@ -395,3 +399,47 @@ func UpdatePodCondition(status *v1.PodStatus, condition *v1.PodCondition) bool {
+// Returns true if pod TotalResourcesRequested has changed.
+func UpdateTotalResources(status *v1.PodStatus, pod *v1.Pod) bool {
+	requests := v1.ResourceList{}
+	// TODO: de-dupe this calculation
```
Will update this PR or https://github.com/kubernetes/kubernetes/pull/115367, depending on which one goes in first.

SergeyKanzhelev commented on pkg/api/v1/resource/helpers.go:
```diff
@@ -24,8 +24,16 @@ import (
+)
+
+// ContainerType signifies container type. This is replicated here to avoid an import cycle.
+type ContainerType int
```
I wonder if we should just declare it in types.go so it can be used universally without worrying of dependencies.

SergeyKanzhelev commented on pkg/apis/core/types.go:
```diff
@@ -3626,6 +3626,13 @@ type PodStatus struct {
+	// and no requests on init or regular containers, the field is omitted.
+	// More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+	// +optional
+	TotalResourcesRequested ResourceList
```
I wonder if TotalRequestedResources will be more aligned with other naming like https://github.com/kubernetes/kubernetes/pull/116450

tzneal replied:
Maybe so? I think @vinaykul suggested this name, any thoughts?

vinaykul replied:
We [recently changed the name](https://github.com/kubernetes/kubernetes/pull/116450) `ResourcesAllocated` to `AllocatedResources` [for consistency](https://kubernetes.slack.com/archives/CJUQN3E4T/p1678415227264449?thread_ts=1678401527.788819&cid=CJUQN3E4T). (Adjective-Noun). So +1 for TotalRequestedResources (sorry!)
/cc @thockin

tzneal replied:
No problem, curious what you think about the semantics.  In the event of a resize that won't work, should it report the requested size or the in-use size?  There's some discussion below with Sergey, I can see utility for both.

vinaykul replied:
Imho it should be what we want (what the pod wants in order to successfully schedule taking into account init-container resources, pod overhead, regular container resources). `Requested` in the name makes that pretty clear.

Adding in-place pod resize to the mix, the signal from the summarized PodStatus.Resize is an easy signal to the user that informs how the `Requested` resize is going. `InProgress` = on the way, "" = done/no resize pending, `Proposed` = ACK we heard you,  `Deferred` = sure but not right now (evict and go elsewhere if you cant wait), `Infeasible` = not gonna happen, evict and go elsewhere to get what you need.

SergeyKanzhelev commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4168,6 +4168,15 @@ type PodStatus struct {
+
+	// TotalResourcesRequested is the effective resource requests for this pod taking into consideration init
+	// and regular containers, and the pod overhead. When the InPlacePodVerticalScaling feature gate is enabled,
+	// the field will show current effective resource request that is applied which may not match the requested value
```
Does this semantic desirable? Will the semantic of what we want be better than what we have now?

tzneal replied:
If it's "what we have", you can use it to determine the amount of free resources on a node where "what we want" isn't helpful for that. The name "TotalResourcesRequested" implies more of "what we want" though.

tzneal replied:
After @vinaykul's comment, rebasing, renaming to `TotalRequestedResources` and changing it be "what we want" which more closely matches the name.

tzneal commented:
/test

tzneal commented:
/retest

bart0sh commented:
/assign @vinaykul

SergeyKanzhelev commented on pkg/api/v1/resource/helpers.go:
```diff
@@ -445,3 +447,34 @@ func MergeContainerResourceLimits(container *v1.Container,
+	})
+
+	if ResourcesEqual(requests, status.TotalRequestedResources) {
+		return false
```
this is the most common code path, correct? Most Pods will only need to have this sent once and will never change. I wonder if there is anything we can do to optimize on this and only update when in-place update happened.

Overhead doesn't seem to be that much though, so should be fine as-is I think

SergeyKanzhelev reviewed (approved):
/lgtm

with the question whether we can improve things

mrunalp commented on pkg/api/v1/pod/util.go:
```diff
@@ -395,3 +395,20 @@ func UpdatePodCondition(status *v1.PodStatus, condition *v1.PodCondition) bool {
 }
+
+// resourcesEqual returns true if the two resources lists provided are equal.
+func resourcesEqual(lhs v1.ResourceList, rhs v1.ResourceList) bool {
```
Can we drop this from the first commit since it is moved to a different file in the second commit?

tzneal replied:
Yup, will do.

SergeyKanzhelev commented:
/assign @ahg-g

ahg-g commented:
/approve

the scheduler bits look good to me

vinaykul commented on pkg/apis/core/types.go:
```diff
@@ -3691,6 +3691,15 @@ type PodStatus struct {
 	ResourceClaimStatuses []PodResourceClaimStatus
+
+	// TotalRequestedResources is the effective resource requests for this pod taking into consideration init
+	// and regular containers, and the pod overhead. When the InPlacePodVerticalScaling feature gate is enabled,
```
To future-proof this comment when InPlacePodVerticalScaling is GA and feature gate is gone, can you please restate this?
Suggestion:
`If a resize of pod resources has been requested for a running pod, this field reflects the total desired resources for that pod which differs from actual resources allocated for the pod by the node.`

vinaykul replied:
@tzneal Besides the above suggestion, it looks good to me.
/lgtm in advance

tzneal replied:
I agree, that's better wording, incorporated.

SergeyKanzhelev commented:
/retest

dims commented:
/approve
/assign @liggitt for API reviews

smarterclayton commented on api/openapi-spec/v3/api__v1_openapi.json:
```diff
@@ -5560,6 +5560,18 @@
+              ],
+              "default": {}
+            },
+            "description": "TotalRequestedResources is the effective resource requests for this pod taking into consideration init and regular containers, and the pod overhead. If a resize of pod resources has been requested for a running pod, this field reflects the total desired resources for that pod which differs from actual resources allocated for the pod by the node.  If there is no effecti ...
```
It's weird to optionally not have this set.  I would probably prefer to always set this, and have it be empty of resources.  Or maybe that's how it works, and the doc is incorrect?

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4298,6 +4298,15 @@ type PodStatus struct {
+	// no requests on init or regular containers, the field is omitted.
+	// More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+	// +optional
+	TotalRequestedResources ResourceList `json:"totalRequestedResources,omitempty" protobuf:"bytes,17,opt,name=totalRequestedResources"`
```
`Total` is a bit redundant.  `pod.status.requestedResources` should already imply pod scope.  Unless we have a very good argument, we should call it `requestedResources`.

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4298,6 +4298,15 @@ type PodStatus struct {
 	// +optional
 	ResourceClaimStatuses []PodResourceClaimStatus `json:"resourceClaimStatuses,omitempty" patchStrategy:"merge,retainKeys" patchMergeKey:"name" protobuf:"bytes,15,rep,name=resourceClaimStatuses"`
+
+	// TotalRequestedResources is the effective resource requests for this pod taking into consideration init
```
Who sets this?  Kubelet?  Scheduler?  Is this field explicitly unset until the pod is scheduled and the node performs an update?  If so, we should describe in this doc when the field is not set.

tzneal replied:
The scheduler sets it upon scheduling failure, the intent there was to allow users to be able to inspect the pod to identify the requested resources since its more difficult to calculate with sidecar init containers.  Otherwise, kubelet owns the value.

tzneal replied:
Modified to be set upon pod creation and then kubelet can update it from there.

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4298,6 +4298,15 @@ type PodStatus struct {
+
+	// TotalRequestedResources is the effective resource requests for this pod taking into consideration init
+	// and regular containers, and the pod overhead. If a resize of pod resources has been requested for a running
+	// pod, this field reflects the total desired resources for that pod which differs from actual resources
```
This sentence is a bit awkward, and reading it I'm not clear what the field does.

Does it:
1. Reflect the current allocated resource state on the kubelet?
2. Reflect the total desired resources for the pod?

Those are two different things, and I can't figure out which one it is.

tzneal replied:
It's 2, I'll updatt the comment to clarify.

smarterclayton commented:
Did a first api review pass.

liggitt commented:
/unassign
/assign @smarterclayton 
for API review

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4298,6 +4298,13 @@ type PodStatus struct {
+
+	// RequestedResources is the effective resource requests for this pod taking into consideration init
+	// and regular containers, and the pod overhead. This field reflects the total desired resources for
+	// the pod which may differs from actual resources allocated for the pod by the node.
```
`may differ`.  

Then one more question - when did we spec that the pod updates this.  After resize is accepted?  Or independent of that, once the kubelet observes the value?  Why?

Depending on that answer, we should add one more sentence: `The node will update this value as it <observes|accepts> changes to requested resources.  If no node can be assigned by the scheduler this value will be set to ease debugging of out-of-resource situations.`

Also, this needs to be feature gated (it's a new field on a v1 object), so should have a comment down below about which feature gate controls it and should only be set when the feature gate is true.

tzneal replied:
> Then one more question - when did we spec that the pod updates this. After resize is accepted? Or independent of that, once the kubelet observes the value? Why?

Its when kubelet observes the value, its called from `generateAPIPodStatus->convertStatusToAPIStatus`.  It looks like resizing the pod does an update of the `.Resize` field in `handlePodResourcesResize`, so I think in the case of sidecar+vertical pod resizing enabled and a pod is being resized, there's a chance it will be out of date for a short period.

vinaykul replied:
@tzneal I have not looked deeply into this besides the parts related to in-place resize. IIRC from KEP discussions a few months ago, am I correct in saying that the primary use case for this is for an entity like cluster autoscaler to easily see the pod resource needs in order to provision additional nodes? If yes, can this not be set (protected by featuregate that @smarterclayton pointed out) in apiserver? (PodStrategy.PrepareForCreate much like QOSClass). IMHO that might simplify things with kubelet being the only writer after the initial write.

tzneal replied:
> am I correct in saying that the primary use case for this is for an entity like cluster autoscaler 

I don't think so, the primary use case is users that want to know why pods won't schedule.  Its a bit difficult to look at a pod spec with a mixture of init containers, sidecar init contianers and main containers and know the actual amount of resources required.

That's a good thought in the PrepareForCreate, I think its cleaner and will look into it.

tzneal commented:
/retest

tzneal commented:
/retest

tzneal commented:
/retest

tzneal commented:
/retest

tzneal commented:
/retest
